File: companion-android/app/src/test/java/me/fiksu/esp32map/companion/integration/ble/FakeRouteSyncBluetoothClient.kt
package me.fiksu.esp32map.companion.integration.ble

import me.fiksu.esp32map.companion.domain.DeviceConnectionState
import me.fiksu.esp32map.companion.domain.RouteSyncFaultInjectionMode
import me.fiksu.esp32map.companion.domain.RouteSyncMessage

/**
 * Test double for [RouteSyncBluetoothClient]. Records what the service
 * called for it and lets the test inject results / errors. Mirrors the
 * shape of `companion-ios/.../FakeRouteSyncBluetoothClient.swift` so the
 * cross-platform pairing-flow tests stay in lockstep.
 */
class FakeRouteSyncBluetoothClient : RouteSyncBluetoothClient {
    override var onSyncMessage: ((RouteSyncMessage) -> Unit)? = null
    override var onConnectionStateChange: ((DeviceConnectionState, String?) -> Unit)? = null

    var isReadyOverride: Boolean = false

    override val isReady: Boolean
        get() = isReadyOverride

    var scanCallCount = 0
        private set
    var connectCallCount = 0
        private set
    var connectToPairedCallCount = 0
        private set
    var writeCallCount = 0
        private set
    var lastWrittenPacket: BleRouteSyncPacket? = null
        private set
    var lastConnectedIdentifier: String? = null
        private set

    var scanResult: Result<String> = Result.success("ESP32 Bike Minimap")
    var connectResult: Result<String> = Result.success("ESP32 Bike Minimap")
    var connectPairedResult: Result<String> = Result.success("ESP32 Bike Minimap")
    var connectAdvertisedResult: Result<String> = Result.success("ESP32 Bike Minimap")
    var writeResult: Result<Unit> = Result.success(Unit)
    var writePairingConfirmResult: Result<Unit> = Result.success(Unit)

    var connectToAdvertisedCallCount = 0
        private set
    var lastAdvertisedIdentifier: String? = null
        private set
    var writePairingConfirmCallCount = 0
        private set
    var lastWrittenPairingSecret: ByteArray? = null
        private set

    override fun armDebugFault(mode: RouteSyncFaultInjectionMode) {
        // No-op: fault injection isn't observable through this fake yet.
    }

    override suspend fun scanForRouteSyncPeripheral(timeoutMs: Long): String {
        scanCallCount += 1
        return scanResult.getOrThrow()
    }

    override suspend fun connectToScannedPeripheral(): String {
        connectCallCount += 1
        return connectResult.getOrThrow()
    }

    override suspend fun connectToPairedPeripheral(identifier: String): String {
        connectToPairedCallCount += 1
        lastConnectedIdentifier = identifier
        return connectPairedResult.getOrThrow()
    }

    override suspend fun connectToAdvertisedPeripheral(identifier: String): String {
        connectToAdvertisedCallCount += 1
        lastAdvertisedIdentifier = identifier
        return connectAdvertisedResult.getOrThrow()
    }

    override suspend fun writePairingConfirm(secret: ByteArray) {
        writePairingConfirmCallCount += 1
        lastWrittenPairingSecret = secret
        writePairingConfirmResult.getOrThrow()
    }

    override suspend fun write(packet: BleRouteSyncPacket) {
        writeCallCount += 1
        lastWrittenPacket = packet
        writeResult.getOrThrow()
    }
}

File: companion-apps/ios/ESP32MapCompanionTests/Fakes/FakeRouteSyncBluetoothClient.swift
import Foundation
@testable import ESP32MapCompanion

/// Test double for `RouteSyncBluetoothClient`. Records call counts and arguments
/// so service- and AppModel-level tests can assert on the BLE traffic without
/// touching CoreBluetooth.
final class FakeRouteSyncBluetoothClient: RouteSyncBluetoothClient {
    struct PhoneGpsWrite: Equatable {
        let lat: Double
        let lon: Double
        let speed: Double
        let course: Double?
        let accuracy: Double?
    }

    var onSyncMessage: ((RouteSyncMessage) -> Void)?
    var onConnectionStateChange: ((DeviceConnectionState, String?) -> Void)?

    var isReady: Bool = false

    private(set) var scanCallCount = 0
    private(set) var connectCallCount = 0
    private(set) var connectToPairedCallCount = 0
    private(set) var connectToAdvertisedCallCount = 0
    private(set) var writeCallCount = 0
    private(set) var writePairingConfirmCallCount = 0
    private(set) var writePairingRequestCallCount = 0
    private(set) var armDebugFaultCallCount = 0

    private(set) var lastScanTimeout: TimeInterval?
    private(set) var lastConnectedIdentifier: String?
    private(set) var lastWrittenPacket: BleRouteSyncPacket?
    private(set) var lastWrittenPairingSecret: Data?
    private(set) var lastArmedDebugFault: RouteSyncFaultInjectionMode?
    private(set) var phoneGpsWrites: [PhoneGpsWrite] = []

    /// What `scanForRouteSyncPeripheral` should produce: a device name on
    /// success or an error to throw. Defaults to a stub name.
    var scanResult: Result<String, Error> = .success("ESP32 Bike Minimap")
    var connectResult: Result<String, Error> = .success("ESP32 Bike Minimap")
    var connectToPairedResult: Result<String, Error> = .success("ESP32 Bike Minimap")
    var connectToAdvertisedResult: Result<ConnectedPeripheralInfo, Error> = .success(
        ConnectedPeripheralInfo(name: "ESP32 Bike Minimap", identifier: "B6C8AE6A-1A8C-4F2E-9F1A-9D2B0C3E4F5A")
    )
    var writeResult: Result<Void, Error> = .success(())
    var writePairingConfirmResult: Result<Void, Error> = .success(())
    var writePairingRequestResult: Result<Void, Error> = .success(())

    func armDebugFault(_ mode: RouteSyncFaultInjectionMode) {
        armDebugFaultCallCount += 1
        lastArmedDebugFault = mode
    }

    func scanForRouteSyncPeripheral(timeout: TimeInterval) async throws -> String {
        scanCallCount += 1
        lastScanTimeout = timeout
        return try scanResult.get()
    }

    func connectToScannedPeripheral() async throws -> String {
        connectCallCount += 1
        return try connectResult.get()
    }

    func connectToPairedPeripheral(identifier: String) async throws -> String {
        connectToPairedCallCount += 1
        lastConnectedIdentifier = identifier
        return try connectToPairedResult.get()
    }

    func connectToAdvertisedPeripheral() async throws -> ConnectedPeripheralInfo {
        connectToAdvertisedCallCount += 1
        let info = try connectToAdvertisedResult.get()
        lastConnectedIdentifier = info.identifier
        return info
    }

    func writePairingConfirm(secret: Data) async throws {
        writePairingConfirmCallCount += 1
        lastWrittenPairingSecret = secret
        try writePairingConfirmResult.get()
    }

    func writePairingRequest() async throws {
        writePairingRequestCallCount += 1
        try writePairingRequestResult.get()
    }

    func write(packet: BleRouteSyncPacket) async throws {
        writeCallCount += 1
        lastWrittenPacket = packet
        try writeResult.get()
    }

    func writePhoneGpsSample(lat: Double, lon: Double, speed: Double, course: Double?, accuracy: Double?) async throws {
        phoneGpsWrites.append(
            PhoneGpsWrite(
                lat: lat,
                lon: lon,
                speed: speed,
                course: course,
                accuracy: accuracy
            )
        )
    }
}
